src/checked_transaction.rs

```diff
@@ -92,6 +104,58 @@ impl CheckedTransaction {
         self.checked_signatures
     }
 
+    pub const fn metadata(&self) -> Option<&Metadata> {
+        self.transaction.metadata()
+    }
+
+    /// Update change and variable outputs.
+    ///
+    /// `revert` will signal if the execution was reverted. It will refund the unused gas cost to
+    /// the base asset and reset output changes to their initial balances.
+    ///
+    /// `unused_gas` will be added to the output that contains the base asset
+    ///
+    /// `balances` will contain the current state of the free balances
+    pub fn update_outputs<I>(&mut self, revert: bool, unused_gas: Word, balances: &I)
+    where
+        I: for<'a> Index<&'a AssetId, Output = Word>,
+    {
+        self.transaction
+            ._outputs_mut()
+            .iter_mut()
+            .for_each(|o| match o {
+                // If revert, set base asset to initial balance and refund unused gas
+                //
+                // Note: the initial balance deducts the gas limit from base asset
+                Output::Change {
+                    asset_id, amount, ..
+                } if revert && asset_id == &BASE_ASSET => {
+                    *amount = self.initial_free_balances[&BASE_ASSET] + unused_gas
```

Comment:
The following is psuedo-code, and is intended to demonstrate the thinking from here: https://github.com/FuelLabs/fuel-tx/pull/162/files#r920644324

```suggestion
                    refund = (self.gas_limit() - unused_gas).checked_mul(gas_price).div_floor(gas_price_factor);
                    *amount = self.initial_free_balances[&BASE_ASSET] + refund
```

Reply:
Updated in the previous comment